packages/core/src/data-access/users.ts

```diff
@@ -0,0 +1,98 @@
+import { AppContext } from '$core/data-access'
+import { getWorkspace } from '$core/data-access/workspaces'
+import { Result, verifyPassword } from '$core/lib'
+import { PromisedResult } from '$core/lib/Transaction'
+import { SafeUser, users } from '$core/schema'
+import { eq } from 'drizzle-orm'
+import { z } from 'zod'
+
+class UserNotFoundError extends Error {
+  constructor() {
+    super('User not found')
+  }
+}
+
+export const loginSchema = z.object({
+  email: z.string().email(),
+  password: z.string(),
+})
+
+export type SessionData = {
+  user: SafeUser
+  workspace: { id: number; name: string }
+}
+
+export async function getUserFromCredentials(
+  credentials: { email: unknown; password: unknown },
```

Comment:
How are these unknown? Aren't they strings?

Reply:
Not anymore